### ISSUE TITLE
ReactiveRedisOperationsSessionRepository.save(…) on changeSessionId fails with ERR no such key

### ISSUE THREAD

**Issue by mp911de** [spring-projects/spring-session#954]
Saving a session from `changeSessionId()`fails with `ERR no such key`

* Repository with code to reproduce: \
  https://github.com/hantsy/spring-reactive-microservice-sample
* Test to reproduce: \
   `com.example.authservice.IntegrationTests` (Please make sure to run Redis and a Apache Cassandra node, does not require further setup beyond install & start)
* Original StackOverflow post: \
   https://stackoverflow.com/questions/47829537/exception-threw-in-integrationtest-rediscommandexecutionexception-err-no-such

Full log:

```
2017-12-15 14:59:16.757  INFO 24400 --- [           main] c.example.authservice.IntegrationTests   : Starting IntegrationTests on Marks-Mac-Pro.fritz.box with PID 24400 (started by mark in /Users/mark/tmp/spring-reactive-microservice-sample)
2017-12-15 14:59:16.759  INFO 24400 --- [           main] c.example.authservice.IntegrationTests   : No active profile set, falling back to default profiles: default
2017-12-15 14:59:16.799  INFO 24400 --- [           main] onfigReactiveWebServerApplicationContext : Refreshing org.springframework.boot.web.reactive.context.AnnotationConfigReactiveWebServerApplicationContext@66ea1466: startup date [Fri Dec 15 14:59:16 CET 2017]; root of context hierarchy
2017-12-15 14:59:17.587  INFO 24400 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Overriding bean definition for bean 'cassandraTemplate' with a different definition: replacing [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=cassandraConfig; factoryMethodName=cassandraTemplate; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [com/example/authservice/CassandraConfig.class]] with [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=org.springframework.boot.autoconfigure.data.cassandra.CassandraDataAutoConfiguration; factoryMethodName=cassandraTemplate; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [org/springframework/boot/autoconfigure/data/cassandra/CassandraDataAutoConfiguration.class]]
2017-12-15 14:59:17.590  INFO 24400 --- [           main] o.s.b.f.s.DefaultListableBeanFactory     : Overriding bean definition for bean 'reactiveCassandraTemplate' with a different definition: replacing [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=cassandraConfig; factoryMethodName=reactiveCassandraTemplate; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [com/example/authservice/CassandraConfig.class]] with [Root bean: class [null]; scope=; abstract=false; lazyInit=false; autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=org.springframework.boot.autoconfigure.data.cassandra.CassandraReactiveDataAutoConfiguration; factoryMethodName=reactiveCassandraTemplate; initMethodName=null; destroyMethodName=(inferred); defined in class path resource [org/springframework/boot/autoconfigure/data/cassandra/CassandraReactiveDataAutoConfiguration.class]]
2017-12-15 14:59:17.602  INFO 24400 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Multiple Spring Data modules found, entering strict repository configuration mode!
2017-12-15 14:59:17.669  INFO 24400 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Multiple Spring Data modules found, entering strict repository configuration mode!
2017-12-15 14:59:17.685  INFO 24400 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Multiple Spring Data modules found, entering strict repository configuration mode!
2017-12-15 14:59:17.701  INFO 24400 --- [           main] .RepositoryConfigurationExtensionSupport : Spring Data Redis - Could not safely identify store assignment for repository candidate interface com.example.authservice.UserRepository.
2017-12-15 14:59:18.443  INFO 24400 --- [           main] o.h.v.i.engine.ValidatorFactoryImpl      : HV000238: Temporal validation tolerance set to 0.
2017-12-15 14:59:18.719  INFO 24400 --- [           main] c.d.driver.core.GuavaCompatibility       : Detected Guava >= 19 in the classpath, using modern compatibility layer
2017-12-15 14:59:18.972  INFO 24400 --- [           main] com.datastax.driver.core.ClockFactory    : Using native clock to generate timestamps.
2017-12-15 14:59:19.156  WARN 24400 --- [           main] com.datastax.driver.core.NettyUtil       : Found Netty's native epoll transport, but not running on linux-based operating system. Using NIO instead.
2017-12-15 14:59:20.662  WARN 24400 --- [           main] com.datastax.driver.core.Cluster         : You listed localhost/0:0:0:0:0:0:0:1:9042 in your contact points, but it wasn't found in the control host's system.peers at startup
2017-12-15 14:59:20.757  WARN 24400 --- [           main] licationStrategy$NetworkTopologyStrategy : Error while computing token map for keyspace full2 with datacenter bar: could not achieve replication factor 4 (found 0 replicas only), check your keyspace replication settings.
2017-12-15 14:59:20.758  WARN 24400 --- [           main] licationStrategy$NetworkTopologyStrategy : Error while computing token map for keyspace full2 with datacenter foo: could not achieve replication factor 4 (found 0 replicas only), check your keyspace replication settings.
2017-12-15 14:59:20.759  INFO 24400 --- [           main] c.d.d.c.p.DCAwareRoundRobinPolicy        : Using data-center name 'datacenter1' for DCAwareRoundRobinPolicy (if this is incorrect, please provide the correct datacenter name with DCAwareRoundRobinPolicy constructor)
2017-12-15 14:59:20.761  INFO 24400 --- [           main] com.datastax.driver.core.Cluster         : New Cassandra host localhost/127.0.0.1:9042 added
2017-12-15 14:59:21.495  WARN 24400 --- [luster-worker-1] licationStrategy$NetworkTopologyStrategy : Error while computing token map for keyspace full2 with datacenter bar: could not achieve replication factor 4 (found 0 replicas only), check your keyspace replication settings.
2017-12-15 14:59:21.495  WARN 24400 --- [luster-worker-1] licationStrategy$NetworkTopologyStrategy : Error while computing token map for keyspace full2 with datacenter foo: could not achieve replication factor 4 (found 0 replicas only), check your keyspace replication settings.
2017-12-15 14:59:22.202 DEBUG 24400 --- [           main] s.w.r.r.m.a.RequestMappingHandlerMapping : Looking for request mappings in application context: org.springframework.boot.web.reactive.context.AnnotationConfigReactiveWebServerApplicationContext@66ea1466: startup date [Fri Dec 15 14:59:16 CET 2017]; root of context hierarchy
2017-12-15 14:59:22.265 DEBUG 24400 --- [           main] o.s.w.r.f.s.s.RouterFunctionMapping      : Looking for router functions in application context: org.springframework.boot.web.reactive.context.AnnotationConfigReactiveWebServerApplicationContext@66ea1466: startup date [Fri Dec 15 14:59:16 CET 2017]; root of context hierarchy
2017-12-15 14:59:22.303  INFO 24400 --- [           main] o.s.w.r.f.s.s.RouterFunctionMapping      : Mapped (GET && /user) -> com.example.authservice.AuthServiceApplication$$Lambda$474/302256620@273293c8
(GET && /users/exists) -> com.example.authservice.AuthServiceApplication$$Lambda$476/1771486148@6f1d799
2017-12-15 14:59:22.320  INFO 24400 --- [           main] o.s.w.r.handler.SimpleUrlHandlerMapping  : Mapped URL path [/webjars/**] onto handler of type [class org.springframework.web.reactive.resource.ResourceWebHandler]
2017-12-15 14:59:22.321  INFO 24400 --- [           main] o.s.w.r.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**] onto handler of type [class org.springframework.web.reactive.resource.ResourceWebHandler]
2017-12-15 14:59:22.363  INFO 24400 --- [           main] o.h.v.i.engine.ValidatorFactoryImpl      : HV000238: Temporal validation tolerance set to 0.
2017-12-15 14:59:22.400  INFO 24400 --- [           main] o.s.w.r.r.m.a.ControllerMethodResolver   : Looking for @ControllerAdvice: org.springframework.boot.web.reactive.context.AnnotationConfigReactiveWebServerApplicationContext@66ea1466: startup date [Fri Dec 15 14:59:16 CET 2017]; root of context hierarchy
2017-12-15 14:59:23.446  INFO 24400 --- [ctor-http-nio-1] r.ipc.netty.tcp.BlockingNettyContext     : Started HttpServer on /0:0:0:0:0:0:0:0:54930
2017-12-15 14:59:23.447  INFO 24400 --- [           main] o.s.b.web.embedded.netty.NettyWebServer  : Netty started on port(s): 54930
2017-12-15 14:59:23.451  INFO 24400 --- [           main] com.example.authservice.DataInitializer  : start users initialization  ...
2017-12-15 14:59:23.471  INFO 24400 --- [           main] reactor.Flux.ConcatArray.1               : onSubscribe(FluxConcatArray.ConcatArraySubscriber)
2017-12-15 14:59:23.472  INFO 24400 --- [           main] reactor.Flux.ConcatArray.1               : request(unbounded)
2017-12-15 14:59:23.478  INFO 24400 --- [           main] c.example.authservice.IntegrationTests   : Started IntegrationTests in 7.204 seconds (JVM running for 8.716)
2017-12-15 14:59:23.819  INFO 24400 --- [er-nio-worker-3] reactor.Flux.ConcatArray.1               : onNext(User(username=admin, password={bcrypt}$2a$10$GWL/MzYrTWKxclm2xECEIO8WHu2t//FFOMKkCPu3/NyE234f6Y/GS, email=admin@example.com, active=true, roles=[USER, ADMIN]))
2017-12-15 14:59:23.819  INFO 24400 --- [er-nio-worker-3] reactor.Flux.ConcatArray.1               : onNext(User(username=user, password={bcrypt}$2a$10$CaX07.68xqlYUeXkeDJouORAc37HYgffjv7hqWzsbwjJiDXVgpgGm, email=user@example.com, active=true, roles=[USER]))
2017-12-15 14:59:23.820  INFO 24400 --- [er-nio-worker-3] reactor.Flux.ConcatArray.1               : onComplete()
2017-12-15 14:59:23.820  INFO 24400 --- [er-nio-worker-3] com.example.authservice.DataInitializer  : done users initialization...
2017-12-15 14:59:23.988 DEBUG 24400 --- [           main] o.s.web.reactive.function.client         : onSubscribe(FluxMap.MapSubscriber)
2017-12-15 14:59:23.988 DEBUG 24400 --- [           main] o.s.web.reactive.function.client         : request(unbounded)
2017-12-15 14:59:24.288 DEBUG 24400 --- [ctor-http-nio-5] o.s.web.reactive.function.client         : onNext(WiretapClientHttpResponse [delegate=ReactorClientHttpResponse{request=[GET /user],status=401}])
2017-12-15 14:59:24.290 DEBUG 24400 --- [ctor-http-nio-5] o.s.web.reactive.function.client         : onComplete()
2017-12-15 14:59:24.311 DEBUG 24400 --- [           main] o.s.web.reactive.function.client         : onSubscribe(FluxMap.MapSubscriber)
2017-12-15 14:59:24.312 DEBUG 24400 --- [           main] o.s.web.reactive.function.client         : request(unbounded)
2017-12-15 14:59:59.936  INFO 24400 --- [     parallel-3] io.lettuce.core.EpollProvider            : Starting without optional epoll library
2017-12-15 14:59:59.938  INFO 24400 --- [     parallel-3] io.lettuce.core.KqueueProvider           : Starting without optional kqueue library
2017-12-15 15:00:00.025 DEBUG 24400 --- [oEventLoop-12-1] o.s.web.reactive.DispatcherHandler       : Processing GET request for [http://localhost:54930/user]
2017-12-15 15:00:00.028 DEBUG 24400 --- [oEventLoop-12-1] o.s.w.r.function.server.RouterFunctions  : Predicate "(GET && /user)" matches against "GET /user"
2017-12-15 15:00:07.012 ERROR 24400 --- [oEventLoop-12-1] o.s.w.s.adapter.HttpWebHandlerAdapter    : Failed to handle request

org.springframework.data.redis.RedisSystemException: Error in execution; nested exception is io.lettuce.core.RedisCommandExecutionException: ERR no such key
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:54) ~[spring-data-redis-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.data.redis.connection.lettuce.LettuceExceptionConverter.convert(LettuceExceptionConverter.java:41) ~[spring-data-redis-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.data.redis.connection.lettuce.LettuceReactiveRedisConnection.lambda$translateException$0(LettuceReactiveRedisConnection.java:267) ~[spring-data-redis-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at reactor.core.publisher.Flux.lambda$onErrorMap$23(Flux.java:5170) ~[reactor-core-3.1.2.RELEASE.jar:3.1.2.RELEASE]
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88) ~[reactor-core-3.1.2.RELEASE.jar:3.1.2.RELEASE]
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onError(MonoFlatMapMany.java:243) ~[reactor-core-3.1.2.RELEASE.jar:3.1.2.RELEASE]
	at org.springframework.security.test.context.support.ReactorContextTestExecutionListener$DelegateTestExecutionListener$SecuritySubContext.onError(ReactorContextTestExecutionListener.java:101) ~[spring-security-test-5.0.0.RELEASE.jar:5.0.0.RELEASE]
	at org.springframework.security.test.context.support.ReactorContextTestExecutionListener$DelegateTestExecutionListener$SecuritySubContext.onError(ReactorContextTestExecutionListener.java:101) ~[spring-security-test-5.0.0.RELEASE.jar:5.0.0.RELEASE]
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120) ~[reactor-core-3.1.2.RELEASE.jar:3.1.2.RELEASE]
	at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:120) ~[reactor-core-3.1.2.RELEASE.jar:3.1.2.RELEASE]
	at reactor.core.publisher.MonoNext$NextSubscriber.onError(MonoNext.java:87) ~[reactor-core-3.1.2.RELEASE.jar:3.1.2.RELEASE]
	at io.lettuce.core.RedisPublisher$State.onError(RedisPublisher.java:655) ~[lettuce-core-5.0.1.RELEASE.jar:na]
	at io.lettuce.core.RedisPublisher$RedisSubscription.onError(RedisPublisher.java:313) ~[lettuce-core-5.0.1.RELEASE.jar:na]
	at io.lettuce.core.RedisPublisher$SubscriptionCommand.onError(RedisPublisher.java:767) ~[lettuce-core-5.0.1.RELEASE.jar:na]
	at io.lettuce.core.RedisPublisher$SubscriptionCommand.complete(RedisPublisher.java:703) ~[lettuce-core-5.0.1.RELEASE.jar:na]
	at io.lettuce.core.protocol.CommandHandler.decode(CommandHandler.java:558) ~[lettuce-core-5.0.1.RELEASE.jar:na]
	at io.lettuce.core.protocol.CommandHandler.channelRead(CommandHandler.java:511) ~[lettuce-core-5.0.1.RELEASE.jar:na]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:138) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459) ~[netty-transport-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) ~[netty-common-4.1.17.Final.jar:4.1.17.Final]
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138) ~[netty-common-4.1.17.Final.jar:4.1.17.Final]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_131]
Caused by: io.lettuce.core.RedisCommandExecutionException: ERR no such key
	... 26 common frames omitted
```

**Comment by hantsy** [comment]
@mp911de  Is there some workaround for this now?

**Comment by vpavic** [comment]
Thanks for the report @mp911de and @hantsy.

I've reproduced the problem by applying the following diff to our [WebFlux sample app](https://github.com/spring-projects/spring-session/tree/master/samples/boot/webflux):

```diff
diff --git a/samples/boot/webflux/spring-session-sample-boot-webflux.gradle b/samples/boot/webflux/spring-session-sample-boot-webflux.gradle
index 11201e2b..5921614d 100644
--- a/samples/boot/webflux/spring-session-sample-boot-webflux.gradle
+++ b/samples/boot/webflux/spring-session-sample-boot-webflux.gradle
@@ -5,6 +5,7 @@ dependencies {
 	compile "org.springframework.boot:spring-boot-starter-webflux"
 	compile "org.springframework.boot:spring-boot-starter-thymeleaf"
 	compile "org.springframework.boot:spring-boot-starter-data-redis"
+	compile "org.springframework.boot:spring-boot-starter-security"
 	compile "org.springframework.boot:spring-boot-devtools"
 	compile 'org.webjars:bootstrap'
 
diff --git a/samples/boot/webflux/src/main/java/sample/HelloWebFluxApplication.java b/samples/boot/webflux/src/main/java/sample/HelloWebFluxApplication.java
index 5d68d873..9a618dba 100644
--- a/samples/boot/webflux/src/main/java/sample/HelloWebFluxApplication.java
+++ b/samples/boot/webflux/src/main/java/sample/HelloWebFluxApplication.java
@@ -18,6 +18,12 @@ package sample;
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import org.springframework.security.config.web.server.ServerHttpSecurity;
+import org.springframework.security.core.userdetails.MapReactiveUserDetailsService;
+import org.springframework.security.core.userdetails.User;
+import org.springframework.security.web.server.SecurityWebFilterChain;
+import org.springframework.security.web.server.context.WebSessionServerSecurityContextRepository;
 
 /**
  * @author Rob Winch
@@ -29,4 +35,26 @@ public class HelloWebFluxApplication {
 		SpringApplication.run(HelloWebFluxApplication.class, args);
 	}
 
+	@Bean
+	public MapReactiveUserDetailsService userDetailsRepository() {
+		return new MapReactiveUserDetailsService(User.withDefaultPasswordEncoder()
+				.username("user").password("password").roles("USER").build());
+	}
+
+	@Bean
+	public SecurityWebFilterChain springWebFilterChain(ServerHttpSecurity http) {
+		// @formatter:off
+		return http
+			.authorizeExchange()
+				.anyExchange().authenticated()
+				.and()
+			.httpBasic()
+				.securityContextRepository(new WebSessionServerSecurityContextRepository())
+				.and()
+			.csrf()
+				.disable()
+			.build();
+		// @formatter:on
+	}
+
 }
```

**Comment by vpavic** [comment]
I've tracked the cause of this to `SpringSessionWebSessionStore.SpringSessionWebSession#changeSessionId` rather than `ReactiveRedisOperationsSessionRepository`.

Specifically, changing the `changeSessionId` not to invoke `save`, and therefore behave similarly to its equivalent from Servlet API world (`SessionRepositoryFilter.SessionRepositoryRequestWrapper#changeSessionId`), fixes the problem. The change is basically this diff:

```diff
diff --git a/spring-session-core/src/main/java/org/springframework/session/web/server/session/SpringSessionWebSessionStore.java b/spring-session-core/src/main/java/org/springframework/session/web/server/session/SpringSessionWebSessionStore.java
index 69ca4546..1b711682 100644
--- a/spring-session-core/src/main/java/org/springframework/session/web/server/session/SpringSessionWebSessionStore.java
+++ b/spring-session-core/src/main/java/org/springframework/session/web/server/session/SpringSessionWebSessionStore.java
@@ -138,7 +138,7 @@ public class SpringSessionWebSessionStore<S extends Session> implements WebSessi
 			return Mono.defer(() -> {
 				this.session
 						.changeSessionId();
-				return save();
+				return Mono.empty();
 			});
 		}
```

That said I'm still not sure `ReactiveRedisOperationsSessionRepository` is without blame as that `ERR no such key` on rename operation doesn't look good.

@rwinch Could you take a look at the `SpringSessionWebSessionStore` related bits of this?

**Comment by mp911de** [comment]
I think there are at least two options how to address the issue:

1. Check before renaming if the key exists: It's the safe method but allows a race if some other process interacts with Redis in between.
2. Handle the actual error (`ERR no such key`) in the way that it's a known outcome and continue with the actual work.

Not sure whether `changeSessionId()` is appropriate here, don't know that code so much.

**Comment by vpavic** [comment]
Thanks for following up @mp911de.

Upon further inspection I was able to come up with integration test that reproduces this problem, and have actually discovered a flaw in logic that handles changes to session id.

The problem occurs when `SessionRepository#save` operation is invoked twice on a newly created session. In such scenario, Redis replace operation is skipped in first invocation of save operation due to the fact it's a new session, but unfortunately the assignment of changed session id was also skipped which caused the subsequent save operations to think that session id was changed and that Redis replace operation was needed.

I've also discovered the same flaw with `RedisOperationsSessionRepository`, but it wasn't exposed due to different behavior of Spring Session's Servlet API related components as explained in [this comment](https://github.com/spring-projects/spring-session/issues/954#issuecomment-355260203). I've opened #962 to address the problem in `RedisOperationsSessionRepository`.

As for potential change of `SessionRepositoryFilter.SessionRepositoryRequestWrapper#changeSessionId` to align it with Servlet API counterpart component, I've opened #963.

Thanks for catching this @hantsy, and thanks @mp911de for reporting this and helping out.
